designs/isolate-cluster-upgrades.md

```diff
@@ -0,0 +1,221 @@
+# Isolate workload clusters from management cluster upgrades
+
+## Introduction
+When a management cluster is upgraded, it's management components are also upgraded. This includes the EKS-A controller manager. Given this controller contains the logic to "convert" the EKS-A to the cluster-api API, any changes to this logic between one EKS-A version to the next one can trigger a cascading effect in the workload clusters after upgrading the management cluster. In some scenarios, this can cause unexpected machine rollouts.
+
+## Tenets
+1. **Simple**. Prioritize (in order): simple for the user, simple to maintain, simple to implement.
+
+## Goals
+1. As a cluster operator, I want to upgrade my management clusters without new machines being rolled out in the workload clusters.
+2. As a cluster operator, I want EKS-A to fully take care of the upgrade process, I don't want to have to manually update the workload clusters before I upgrade the management cluster.
+
+## Scope
+### In
+* Design a system that prevents, working transparently to the user, new machines from being rolled out in workload clusters when the management cluster is upgraded.
+### Out
+* Decouple management components update from cluster components upgrade in management clusters.
+
+## Solution
+Use the Kubernetes concepts of `generation` and `observedGeneration` to determine when an cluster configuration has changed and only then, reconcile the state of the world to the desired state. This prevents workload clusters from being affected by a management cluster upgrade: if a workload cluster config has already been reconciled and its `generation` matches its `observedGeneration`, no changes will be applied to the CAPI definition of the cluster unless the EKS-A cluster config changes again, regardless if the CAPI cluster definition generated for the same EKS-A cluster config has changed between controller versions.
+
+### `generation` and `observedGeneration`
+* [`metadata.generation`](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata): a sequence number representing a specific generation of the desired state. Set by the system and monotonically increasing, per-resource. May be compared, such as for RAW and WAW consistency.
```

Comment:
Curious, can a user modify this "generation" or "observedGeneration" field explicitly?

Reply:
Not sure about generation, since that's managed by kubernetes. But it's possible.

They can definitely update `observedGeneration` and the controller will revert it.